File: C/ping-pong/ping-pong.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char** argv) {
  MPI_Init(NULL, NULL);
  int world_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  int world_size;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);

  // We are assuming at least 2 processes for this task
  if (world_size < 2) {
    fprintf(stderr, "World size must be greater than 1 for %s\n", argv[0]);
    MPI_Abort(MPI_COMM_WORLD, 1); 
  }
  
  if (argc != 3) {
    fprintf(stderr, "correct usage:\nping_pong_c [number_of_sends] [bytes_per_send]\n");
    MPI_Abort(MPI_COMM_WORLD, 1); 
  }
  
  int number_of_sends = atoi(argv[1]);
  int bytes_per_send = atoi(argv[2]);
  char* data = (char*)malloc(bytes_per_send);
  
  double time = MPI_Wtime();
  
  if (world_rank == 0) {
    // If we are rank 0, set the number to -1 and send it to process 1
    int i = 0;
    for(i = 0; i<number_of_sends; ++i){
      MPI_Send(&data, bytes_per_send, MPI_CHAR, 1, 0, MPI_COMM_WORLD);
      MPI_Recv(&data, bytes_per_send, MPI_CHAR, 1, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    }
    time = MPI_Wtime() - time;
    MPI_Barrier(MPI_COMM_WORLD);
    double capacity=(bytes_per_send*number_of_sends*8)/(1000000.0*time*MPI_Wtick());
    printf("%lf Mbit/s capacity\n%lfs of delay on single message\n", capacity, (time*MPI_Wtick()/(float)number_of_sends));
  
  } else if (world_rank == 1) {
    int i = 0;
    for(i = 0; i<number_of_sends; ++i){
      MPI_Recv(&data, bytes_per_send, MPI_CHAR, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
      MPI_Send(&data, bytes_per_send, MPI_CHAR, 0, 0, MPI_COMM_WORLD);
    }
  }
  
  MPI_Finalize();
} 

File: Python/ping-pong/ping_pong.py
#!/usr/bin/env python
from mpi4py import MPI
from sys import argv
import time as t
comm = MPI.COMM_WORLD
rank = comm.Get_rank()

#usage:
#ping_pong.py [number_of_sends] [bytes_per_send] [-v]
	
if len(argv)>3 and argv[3]=='-v':
	verbose=True
else:
	verbose=False
if len(argv)>2:
	bytes_per_send=int(argv[2])
else:
	bytes_per_send=16
if len(argv)>1:
	number_of_sends=int(argv[1])
else:
	number_of_sends=25
time=MPI.Wtime()
if rank == 0:
	for i in range(number_of_sends):
   		data = bytearray(bytes_per_send)
  		comm.send(data, dest=1)
		if verbose:
			print "Process 0 sent data"
		data = comm.recv(source=1)
		if verbose:
			print "Process 0 received data"
elif rank == 1:
	for i in range(number_of_sends):
   		data = bytearray(bytes_per_send)
  		data=comm.recv(source=0)
		if verbose:
			print "Process 1 sent data"
		comm.send(data,dest=0)
		if verbose:
			print "Process 1 received data"
else:
	print "Expected two nodes"
#substract the times
time=MPI.Wtime()-time
comm.Barrier()
#perform scaling
capacity=float(bytes_per_send*number_of_sends*8)/(float(1000000)*time)
#at this time 'time' is of unit Mbit/s
print str(capacity)+"Mbit/s capacity "
print str(time/float(number_of_sends))+"s of delay on single message"
if rank==0:
	file=open('./output/results.txt','a')
	file.write(str(bytes_per_send)+' '+str(capacity)+' '+str(time/float(number_of_sends))+'\n')
#time=MPI.Wtime()
#t.sleep(1)
#time=MPI.Wtime()-time
#print time
